### PR TITLE
Make Numba pickle compatible with Ray.

### PR DESCRIPTION
**Pull request by DrTodd13**
@sklam Please have a look at this PR.  It is work-in-progress but would like your feedback.  I added some debug code in serialization.py to help me find where the unpickable generators are.  It turns out that StencilFunc was storing typingctx and targetctx and those contain generators.  In this PR, I don't store those in StencilFunc anymore but get them each time I need them in StencilFunc methods.  That seems to have eliminated the previous pickling error I was getting and now I'm not getting any pickling errors at all.  That is not to say pickling is working correctly...just no obvious errors...see below.

What I'm getting now is it looks like Ray is trying to run a Numba function in another process, presumably after unpickling, that invokes a Numba stencil.  star2 is the Numba stencil function.

```  File "python/ray/_raylet.pyx", line 463, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 415, in ray._raylet.execute_task.function_executor
  File "./todd-stencil.py", line 197, in run_step
    do_star2(A,B,W)
  File "/mnt/home/taanders/numba/intel_numba/numba/numba/core/dispatcher.py", line 420, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/mnt/home/taanders/numba/intel_numba/numba/numba/core/dispatcher.py", line 361, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Untyped global name 'star2': Cannot determine Numba type of <class 'numba.stencils.stencil.StencilFunc'>

File "todd-stencil.py", line 96:
def do_star2(A,B,W):
    <source elided>
    #star2(A,W, out=B)
    B += star2(A,W)
```

### CODE REVIEW

**Comment by sklam** [comment]
We'll need to talk about taking `cloudpickle` as a depedency in the dev meeting.

**Comment by sklam** [comment]
> It turns out that StencilFunc was storing typingctx and targetctx and those contain generators. In this PR, I don't store those in StencilFunc anymore but get them each time I need them in StencilFunc methods.

The changes make sense. Code should not serialize the compiler states (the target/typing contexts).

File: numba/core/serialize.py [hunk 1]
```diff
@@ -159,10 +160,80 @@ def _numba_unpickle(address, bytedata, hashed):
...
 
 
-def dumps(obj):
+from collections import Mapping, Set, Sequence
+
+try:
+    from six import string_types, iteritems
+except ImportError:
+    string_types = (str, unicode) if str is bytes else (str, bytes)
+    iteritems = lambda mapping: getattr(mapping, 'iteritems', mapping.items)()
```
Comment (eric-wieser) [inline numba/core/serialize.py:169]: Python 2 is no longer supported, right? So there's no need to do this.

Comment (DrTodd13) [reply numba/core/serialize.py:169]: That's just some debugging code to find unpickable objects.  That will be removed once we have this working and don't need to track down pickling errors again.

**Comment by stuartarchibald** [comment]
@DrTodd13 as `cloudpickle` is now vendored into Numba and used for general purpose serialization, could this PR perhaps be closed?

**Comment by stuartarchibald** [comment]
> @DrTodd13 as `cloudpickle` is now vendored into Numba and used for general purpose serialization, could this PR perhaps be closed?

CC @DrTodd13.

**Pull request closed without merging**